Introduce types and clients for Metrics APIs

DirectXMan12 opened the pull request:
This PR introduces types and clients for both the custom metrics API and the resource metrics API.

The resource metrics API uses autogenerated clients, while the custom metrics API, due the nature of it's URLs, uses a custom client.

DirectXMan12 commented:
@piosz PTAL

DirectXMan12 commented:
A couple of notes on this PR:

Generating the generated files requires https://github.com/kubernetes/repo-infra/pull/7 (and needs to be placed in vendor).

I've tested this PR against my WIP copy of HPA v2 part III (use the custom metrics API, and also use the actual resource metrics API clients), so the code is at least sane enough to work properly (the bits required for the testing can be found at https://github.com/DirectXMan12/kubernetes/tree/feature/hpa-v2-use-cm-api and https://github.com/DirectXMan12/heapster/tree/feature/use-cm-types, which should turn into actual PRs shortly).

@kubernetes/sig-api-machinery-misc  I wasn't sure how to avoid depending on `k8s.io/client-go/pkg/api` for the autogenerated clients -- a bunch of that autogenerated code seems to rely on api.{Scheme,etc} being a global.  Any ideas there?

piosz commented:
Can we split this PR into 2 separate ones: with custom metrics and metrics? Overall looks good for me, thought I'd like some from api-machinery team to take a look.

DirectXMan12 commented:
cc @kubernetes/sig-api-machinery-pr-reviews PTAL for general sanity

deads2k commented:
> @kubernetes/sig-api-machinery-misc I wasn't sure how to avoid depending on k8s.io/client-go/pkg/api for the autogenerated clients -- a bunch of that autogenerated code seems to rely on api.{Scheme,etc} being a global. Any ideas there?

We have https://github.com/kubernetes/kubernetes/pull/41486 open and lgtm'ed.  Google US is out today, so we'll given Daniel a chance to take a look tomorrow.

DirectXMan12 commented:
ack, I'll pull that into kubernetes/repo-infra#7 once it's merged, and then regen this PR

DirectXMan12 commented:
This should be all set with regards to kubernetes/kubernetes#41486 -- I've regenerated with that code.

DirectXMan12 commented:
After a quick glance-over from someone on @kubernetes/sig-api-machinery-pr-reviews, I'd like to get this merged so that the Kubernetes PRs depending on it can move forward

deads2k commented on pkg/apis/custom_metrics/register.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+// GroupName is the group name use in this package
+const GroupName = "custom-metrics"
```
group names should be domain qualified.  This one looks like `custom-metrics.metrics.k8s.io` based on the repo name.

deads2k commented on pkg/apis/custom_metrics/doc.go:
```diff
@@ -0,0 +1,17 @@
+
+// +k8s:deepcopy-gen=package,register
+
+package custom_metrics // import "k8s.io/metrics/pkg/apis/custom_metrics"
```
You'll need to specify a group name.  See https://github.com/kubernetes/sample-apiserver/blob/master/pkg/apis/wardle/doc.go#L20 as an example.

DirectXMan12 replied:
doesn't that tag only apply for generated clients?  None-the-less, you're right, I should probably have it here.

deads2k commented on pkg/apis/custom_metrics/v1alpha1/register.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+var (
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
```
you need to add your conversions here.

deads2k replied:
RegisterConversions

deads2k replied:
> you need to add your conversions here.

nm, it seems those are auto-wired now.

deads2k commented on pkg/apis/metrics/register.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// GroupName is the group name use in this package
+const GroupName = "metrics"
```
domain qualify.  "metrics.k8s.io" is legal for you if you want it.

deads2k replied:
> domain qualify. "metrics.k8s.io" is legal for you if you want it.

But it could get *really* confusing when deciding if you're referring to a type or a group.

DirectXMan12 replied:
This is a pre-existing API, so it probably shouldn't change names, right?

deads2k replied:
> This is a pre-existing API, so it probably shouldn't change names, right?

What version was your API?  At an alpha or beta level, I'd consider cross registering and sunset the old.

DirectXMan12 replied:
it was alpha

deads2k commented on pkg/apis/metrics/types.go:
```diff
@@ -0,0 +1,86 @@
+
+// +genclient=true
+// +resourceName=nodes
+// +readonly=true
```
Don't you need a "nonNamespaced" tag like here: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L2862 ?

Also, care to open an issue for us to add an example of this to the sample-apiserver?

DirectXMan12 replied:
Yep, missed that tag (it wasn't present in the original types in Heapster).  As for the `resourceName` and `readonly` tags, I have to submit a PR to get them into Kube's client-gen first probably, but yeah, I'll open an issue.

deads2k commented on pkg/apis/metrics/types.go:
```diff
@@ -0,0 +1,86 @@
+	// Container name corresponding to the one from pod.spec.containers.
+	Name string
+	// The memory usage is the memory working set.
+	Usage api.ResourceList
```
Since you've done this, you'll need to be sure that you register the ResourceList and all transitive conversions and deep copies in the appropriate versions.

DirectXMan12 replied:
ResourceList doesn't have a deepcopy, Resource has deepcopy as a method, ResourceList's conversion is marked copy-only, as is Resource's, which AFAICT, means it's not needed for nested conversions.

deads2k commented:
You need fuzzer tests.  See https://github.com/kubernetes/sample-apiserver/blob/master/pkg/apis/wardle/install/install_test.go

DirectXMan12 commented:
Alright, I've addressed all the comments here, I believe (with the exception of cross-registering `metrics`, which we'll deal with later).  I'm going to merge this so the dependent PRs can move forwards.
